### PR TITLE
save failed mpeg7 signature and segment data for fast verification

### PR DESCRIPTION
**Pull request by oscar-davids**
**What does this pull request do? Explain your changes. (required)**

Related to this [issue](https://github.com/livepeer/internal-project-tracking/issues/352)

**Specific updates (required)**

- 
- 
- 

**How did you test each of these updates (required)**

**Does this pull request close any open issues?**

**Checklist:**

- [X] Read the [contribution guide](./doc/contributing.md)
- [X] `make` runs successfully
- [ ] All tests in `./test.sh` pass
- [ ] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

### CODE REVIEW

File: drivers/gssave.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	"net/url"
+	"os"
+	"strconv"
+
+	"cloud.google.com/go/storage"
+	"github.com/golang/glog"
+	"google.golang.org/api/option"
+)
+
+// const fullURL = "https://console.cloud.google.com/storage/browser/fastverifyfaildata/media_b452968_3.ts"
```
Comment (cyberj0g) [inline drivers/gssave.go]: Do we need this comment?

File: drivers/gssave.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	return credsJSON != ""
+}
+
+// SetCreds ...
+func SetCreds(bucket, creds string) {
+	FailSaveBucketName = bucket
+
+	info, err := os.Stat(creds)
+	glog.Infof("bucket %s creds %s is not ex %v is dir %v", bucket, creds, os.IsNotExist(err), info != nil && info.IsDir())
+	if info != nil && !os.IsNotExist(err) && !info.IsDir() {
```
Comment (cyberj0g) [inline drivers/gssave.go]: Maybe just `err==nil && !info.IsDir()`?

File: drivers/gssave.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+func SavePairData2GS(trusturi string, data1 []byte, untrusturi string, data2 []byte, suffix string) error {
+
+	trustpurl, _ := url.Parse(trusturi)
+	untrustpurl, _ := url.Parse(untrusturi)
+	pairstr := strconv.Itoa(rand.Int()) + "-"
+	fileName1 := pairstr + trustpurl.Host + "-trust-" + suffix
+	fileName2 := pairstr + untrustpurl.Host + "-untrust-" + suffix
+	fu, err := Save2GS(fileName1, data1)
```
Comment (cyberj0g) [inline drivers/gssave.go]: Maybe make save operation async? It may have significant latency.

File: server/broadcast.go [hunk 4]
```diff
@@ -559,12 +559,26 @@ func (bsm *BroadcastSessionsManager) chooseResults(ctx context.Context, submitRe
...
-			clog.Infof(ctx, "Video comparison from url=%s and url=%s are equal=%v",
+			clog.Infof(ctx, "Video comparison from url=%s and url=%s are equal=%v saveenable=%v",
 				trustedResult.TranscodeResult.Segments[segmToCheckIndex].Url,
-				untrustedResult.TranscodeResult.Segments[segmToCheckIndex].Url, vequal)
+				untrustedResult.TranscodeResult.Segments[segmToCheckIndex].Url, vequal, drivers.FailSaveEnabled())
+		} else {
+			if drivers.FailSaveEnabled() {
+				if drivers.FailSaveEnabled() {
+					drivers.SavePairData2GS(trustedResult.TranscodeResult.Segments[segmToCheckIndex].Url, trustedHash,
+						untrustedResult.TranscodeResult.Segments[segmToCheckIndex].Url, untrustedHash, "phase1.hash")
```
Comment (cyberj0g) [inline server/broadcast.go]: Does `phase1.hash` have any special meaning?

**Review by cyberj0g — CHANGES_REQUESTED** [review]
LGTM overall, left a few minor comments. +1 for nicely-sized PR.

PS: `TestParseAcceptsEitherDatadirCasing` is failing, is it related to changes?

File: server/broadcast.go (hunk 4 above)
Comment (oscar-davids) [reply server/broadcast.go]: Yeah, that's mean mpeg7 p-hash in phase 2 of fast verification steps.

File: drivers/gssave.go (hunk 1 above)
Comment (oscar-davids) [reply drivers/gssave.go]: Removed this comment

File: drivers/gssave.go (hunk 2 above)
Comment (oscar-davids) [reply drivers/gssave.go]: Fixed!

File: drivers/gssave.go (hunk 3 above)
Comment (oscar-davids) [reply drivers/gssave.go]: Fixed.

**Pull request closed without merging**